src/cascadia/TerminalConnection/ConptyConnection.cpp

```diff
@@ -564,43 +536,31 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
     void ConptyConnection::Close() noexcept
     try
     {
-        const bool isClosing = _transitionToState(ConnectionState::Closing);
-        if (isClosing || _isStateAtOrBeyond(ConnectionState::Closed))
-        {
-            // EXIT POINT
-
-            // _clientExitWait holds a CreateThreadpoolWait() which holds a weak reference to "this".
-            // This manual reset() ensures we wait for it to be teared down via WaitForThreadpoolWaitCallbacks().
-            _clientExitWait.reset();
-
-            _hPC.reset(); // tear down the pseudoconsole (this is like clicking X on a console window)
-
-            // CloseHandle() on pipes blocks until any current WriteFile()/ReadFile() has returned.
-            // CancelSynchronousIo prevents us from deadlocking ourselves.
-            // At this point in Close(), _inPipe won't be used anymore since the UI parts are torn down.
-            // _outPipe is probably still stuck in ReadFile() and might currently be written to.
-            if (_hOutputThread)
-            {
-                CancelSynchronousIo(_hOutputThread.get());
-            }
-
-            _inPipe.reset(); // break the pipes
-            _outPipe.reset();
-
-            if (_hOutputThread)
-            {
-                // Waiting for the output thread to exit ensures that all pending _TerminalOutputHandlers()
-                // calls have returned and won't notify our caller (ControlCore) anymore. This ensures that
-                // we don't call a destroyed event handler asynchronously from a background thread (GH#13880).
-                LOG_LAST_ERROR_IF(WAIT_FAILED == WaitForSingleObject(_hOutputThread.get(), INFINITE));
-                _hOutputThread.reset();
-            }
+        // .reset()ing either of these two will signal ConPTY to send out a CTRL_CLOSE_EVENT to all attached clients.
+        // FYI: The other members of this class are concurrently read by the _hOutputThread
+        // thread running in the background and so they're not safe to be .reset().
+        _hPC.reset();
+        _inPipe.reset();
+
+        _transitionToState(ConnectionState::Closing);
```

Comment:
we used to check whether this was true because it would let us perform closing actions _only once_. Are these remaining actions safe to do multiple times?

Reply:
Actually it's the other way around: We used to potentially perform these actions multiple times and now we don't. This is because the old ran past `_transitionToState()` in `ConptyConnection::_ClientTerminated()` without checking the return value. So it might try to extract the `_hOutputThread` and wait on it, even though the main thread might be concurrently running `ConptyConnection::Close()` and trying to do the exact same thing.

The new code is more robust, because the background thread(s) don't close any system objects anymore. Only the main thread does. Technically we don't need `_transitionToState()` anymore. The fact that the pipe returned `ERROR_BROKEN_PIPE` is indication enough that we may exit.

Actually now that I said that I realize that our `CancelSynchronousIo()` code is a bit racy. If we call it when the output thread is currently not stuck in `ReadFile()` it might miss the signal to exit, which is problematic because `Close()` blocks the UI thread.

Reply:
Fixed in my latest commit!